### PR TITLE
Asymmetric PnP Architecture with AP3P Implementation

### PR DESCRIPTION
**Pull request by AdarshGuptaa**
## 📝 Description
This PR introduces a high-performance Algebraic Perspective-Three-Point (AP3P) solver and overhauls the PnP estimation architecture to utilize an asymmetric design, significantly improving both accuracy and execution speed for camera pose estimation.

## Key changes:
1.  AP3P Integration: Implemented the AP3P algorithm (Ke & Roumeliotis, 2017) in `crates/kornia-3d/src/pnp/ap3p.rs.` This includes ported algebraic root finders (degrees 2, 3, and 4) and robust cheirality filtering.
2. Asymmetric PnP Strategy:
- Hypothesis Generation: Uses AP3P (minimal 3-point solver) to maximize inlier recall during RANSAC.
- Local Optimization (LO): Delegates refit phases to EPnP (overdetermined) to leverage full inlier sets without truncation.
3. Performance Optimizations:
- Eliminated heap allocations in the minimal fit() inner loop using stack-allocated fixed-size arrays.
- Pre-extracted camera intrinsics ($f_x, f_y, c_x, c_y$) to avoid repeated matrix destructuring.
- Disabled LM refinement during minimal fitting to reduce per-iteration latency.

**Fixes/Relates to:** #929

---

## 🧪 How Was This Tested?
- [x] **Unit Tests:** (List new/updated tests)
- [x] **Manual Verification:** (Describe the steps you took)
- [x] **Performance/Edge Cases:** (How does this handle nulls, large data, etc.?)

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [ ] 🟢 **No AI used.**
- [x] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [ ] I am assigned to the linked issue (required before PR submission)
- [x] The linked issue has been approved by a maintainer
- [x] This PR strictly implements what the linked issue describes (no scope creep)
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] (Optional) I have attached screenshots/recordings for UI changes.

---

## 💭 Bench Mark
**Against the openCV pnp algorithm**
```
Scene: 600 points, noise=1.5px, threshold=6.0px, max_iters=10000
══════════════════════════════════════════════════════════════════════════
Solver              Ratio   Inliers/Tot  Recall   R_err°    t_err       ms
──────────────────────────────────────────────────────────────────────────
k_ap3p_lo0            10%      57.8/600    0.96    0.617   0.0888     22.6
k_ap3p_lo1            10%      57.8/600    0.96    0.617   0.0888     22.7
k_ap3p_lo2            10%      57.8/600    0.96    0.617   0.0888     22.7
k_ap3p_lo3            10%      57.8/600    0.96    0.617   0.0888     22.6
k_epnp_lo0            10%      14.0/600    0.23   66.354   3.1113    153.6
k_epnp_lo1            10%      14.0/600    0.23   66.354   3.1113    154.3
k_epnp_lo2            10%      14.0/600    0.23   66.354   3.1113    153.8
k_epnp_lo3            10%      14.0/600    0.23   66.354   3.1113    153.8
opencv_ap3p           10%      58.0/600    0.97    0.217   0.0182    206.1
opencv_epnp           10%       0.0/600    0.00  180.000      inf    394.2
──────────────────────────────────────────────────────────────────────────
k_ap3p_lo0            20%     118.6/600    0.99    0.871   0.0684      2.7
k_ap3p_lo1            20%     118.6/600    0.99    0.871   0.0684      2.8
k_ap3p_lo2            20%     118.6/600    0.99    0.871   0.0684      2.8
k_ap3p_lo3            20%     118.6/600    0.99    0.871   0.0684      2.8
k_epnp_lo0            20%     117.6/600    0.98    0.761   0.0627     73.3
k_epnp_lo1            20%     117.6/600    0.98    0.761   0.0627     72.8
k_epnp_lo2            20%     117.6/600    0.98    0.761   0.0627     72.9
k_epnp_lo3            20%     117.6/600    0.98    0.761   0.0627     73.2
opencv_ap3p           20%     120.2/600    1.00    0.124   0.0123     89.1
opencv_epnp           20%     118.8/600    0.99    0.108   0.0100    390.8
──────────────────────────────────────────────────────────────────────────
k_ap3p_lo0            30%     169.6/600    0.94    1.039   0.0879      1.0
k_ap3p_lo1            30%     169.6/600    0.94    1.039   0.0879      1.0
k_ap3p_lo2            30%     169.6/600    0.94    1.039   0.0879      1.0
k_ap3p_lo3            30%     169.6/600    0.94    1.039   0.0879      1.0
k_epnp_lo0            30%     172.8/600    0.96    0.561   0.0644     17.1
k_epnp_lo1            30%     172.8/600    0.96    0.561   0.0644     17.1
k_epnp_lo2            30%     172.8/600    0.96    0.561   0.0644     17.2
k_epnp_lo3            30%     172.8/600    0.96    0.561   0.0644     17.0
opencv_ap3p           30%     177.2/600    0.98    0.095   0.0101     19.1
opencv_epnp           30%     179.6/600    1.00    0.094   0.0076    112.6
──────────────────────────────────────────────────────────────────────────
k_ap3p_lo0            50%     296.6/600    0.99    0.645   0.0467      0.2
k_ap3p_lo1            50%     296.6/600    0.99    0.645   0.0467      0.3
k_ap3p_lo2            50%     296.6/600    0.99    0.645   0.0467      0.3
k_ap3p_lo3            50%     296.6/600    0.99    0.645   0.0467      0.2
k_epnp_lo0            50%     292.6/600    0.98    0.787   0.0643      2.9
k_epnp_lo1            50%     292.6/600    0.98    0.787   0.0643      3.0
k_epnp_lo2            50%     292.6/600    0.98    0.787   0.0643      2.9
k_epnp_lo3            50%     292.6/600    0.98    0.787   0.0643      2.9
opencv_ap3p           50%     295.4/600    0.98    0.060   0.0066      2.6
opencv_epnp           50%     299.2/600    1.00    0.068   0.0074      8.9
══════════════════════════════════════════════════════════════════════════
```

### CODE REVIEW

**Comment by AdarshGuptaa** [comment]
I will accept and make changes based on the reviews and then clean up the comments and docs in the next few commits. 
Any approval will be appreciated @edgarriba

**Comment by edgarriba** [comment]
Some errors still quite large compared to opencv. Make sure also that the kpts generation follows some kind of real data distribution

**Comment by AdarshGuptaa** [comment]
```
══════════════════════════════════════════════════════════════════════════
Solver              Ratio   Inliers/Tot  Recall   R_err°    t_err       ms
──────────────────────────────────────────────────────────────────────────
k_ap3p_lo0            10%      61.0/600    1.02    0.268   0.0276     29.9
k_ap3p_lo1            10%      60.8/600    1.01    0.253   0.0294     29.7
k_ap3p_lo2            10%      61.2/600    1.02    0.242   0.0188     30.1
k_ap3p_lo3            10%      60.8/600    1.01    0.238   0.0324     29.2
k_epnp_lo0            10%      56.4/600    0.94    1.284   0.1254    160.9
k_epnp_lo1            10%      61.0/600    1.02    0.241   0.0211    162.3
k_epnp_lo2            10%      61.0/600    1.02    0.239   0.0224    162.0
k_epnp_lo3            10%      56.6/600    0.94    1.258   0.1259    161.0
opencv_ap3p           10%      60.6/600    1.01    0.315   0.0268    208.1
opencv_epnp           10%      30.6/600    0.51   12.450   1.1784    393.4
──────────────────────────────────────────────────────────────────────────
k_ap3p_lo0            20%     120.6/600    1.01    0.227   0.0278      6.5
k_ap3p_lo1            20%     120.6/600    1.01    0.233   0.0320      7.8
k_ap3p_lo2            20%     120.6/600    1.01    0.233   0.0320      7.1
k_ap3p_lo3            20%     120.6/600    1.01    0.238   0.0329      6.8
k_epnp_lo0            20%     120.6/600    1.01    0.234   0.0259     91.0
k_epnp_lo1            20%     120.6/600    1.01    0.270   0.0287     72.1
k_epnp_lo2            20%     120.6/600    1.01    0.243   0.0287     73.8
k_epnp_lo3            20%     120.6/600    1.01    0.271   0.0300     75.9
opencv_ap3p           20%     119.6/600    1.00    0.286   0.0292     91.1
opencv_epnp           20%     118.6/600    0.99    0.280   0.0286    389.2
──────────────────────────────────────────────────────────────────────────
k_ap3p_lo0            30%     180.4/600    1.00    0.111   0.0121      4.7
k_ap3p_lo1            30%     180.4/600    1.00    0.129   0.0129      7.2
k_ap3p_lo2            30%     180.4/600    1.00    0.117   0.0138      5.9
k_ap3p_lo3            30%     180.4/600    1.00    0.122   0.0127      6.5
k_epnp_lo0            30%     180.4/600    1.00    0.089   0.0107     18.1
k_epnp_lo1            30%     180.4/600    1.00    0.124   0.0135     20.2
k_epnp_lo2            30%     180.4/600    1.00    0.116   0.0130     19.4
k_epnp_lo3            30%     180.4/600    1.00    0.127   0.0138     19.1
opencv_ap3p           30%     178.4/600    0.99    0.188   0.0155     18.8
opencv_epnp           30%     179.2/600    1.00    0.128   0.0138    113.4
──────────────────────────────────────────────────────────────────────────
k_ap3p_lo0            50%     299.4/600    1.00    0.482   0.0439      6.7
k_ap3p_lo1            50%     299.4/600    1.00    0.488   0.0444     11.6
k_ap3p_lo2            50%     299.4/600    1.00    0.483   0.0440      9.6
k_ap3p_lo3            50%     299.4/600    1.00    0.488   0.0444      9.7
k_epnp_lo0            50%     299.4/600    1.00    0.689   0.0652      8.4
k_epnp_lo1            50%     299.4/600    1.00    0.720   0.0697     13.2
k_epnp_lo2            50%     299.4/600    1.00    0.714   0.0693     11.0
k_epnp_lo3            50%     299.4/600    1.00    0.709   0.0688     11.0
opencv_ap3p           50%     297.6/600    0.99    0.237   0.0203      2.5
opencv_epnp           50%     299.0/600    1.00    0.386   0.0364      8.9
══════════════════════════════════════════════════════════════════════════
```

New Benchmarks:

Reasoning behind anomalies:

- 50% ratio gap: f32 floating-point precision degrades when optimizing large datasets (like 300 inliers), creating an error floor that OpenCV avoids by using f64.
- SPRT speed Gap: OpenCV uses a statistical early-exit strategy (SPRT) to instantly terminate the RANSAC loop when a massive inlier set is found, skipping unnecessary iterations for large Inlier ratios.

**Comment by edgarriba** [comment]
for the numbers seems that the more inliers are included, the biggest is the ration error compared to opencv

File: crates/kornia-3d/src/ransac/estimators/ap3p.rs [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+/// Carries the camera intrinsics, optional lens distortion, and solver hyperparameters
+/// as state. Intrinsics are pre-extracted into scalar fields upon initialization to
+/// completely eliminate matrix destructuring overhead during the millions of
+/// `residual()` calls made by the RANSAC inner loop.
+#[derive(Debug, Clone)]
+pub struct AP3PEstimator {
+    /// Full intrinsics matrix (f32, SIMD-aligned).
+    pub k: Mat3AF32,
+    /// Pre-extracted focal length X to bypass inner-loop matrix copies.
+    fx: f32,
```
Comment (edgarriba) [inline crates/kornia-3d/src/ransac/estimators/ap3p.rs]: should we keep all public to be consistent?

File: crates/kornia-3d/src/ransac/estimators/epnp.rs
Comment (edgarriba) [inline crates/kornia-3d/src/ransac/estimators/epnp.rs:1]: @cjpurackal for visibility in changes for epnp

**Comment by AdarshGuptaa** [comment]
> for the numbers seems that the more inliers are included, the biggest is the ration error compared to opencv

Currently Kornia checks all the points to score a model which is brute-force. OpenCV uses **Sequential Probability Ratio Test** or **SPRT** to reject a sufficiently bad model early on. Thats why for larger ratios Kornia seems to fall behind opeCV errors.

**SPRT** could be implemented as a seperate PR shortly after this.

File: crates/kornia-3d/src/ransac/estimators/ap3p.rs [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    type Model = AP3PModel;
+    type Sample = Match2d3d;
+
+    /// AP3P is a minimal solver requiring exactly 3 geometric points.
+    const SAMPLE_SIZE: usize = 3;
+
+    /// Generates camera pose hypotheses from a minimal 3-point sample.
+    ///
+    /// This method is highly optimized for the RANSAC inner loop. It guarantees
+    /// zero heap allocations (`Vec::new`, etc.) and yields up to 4 algebraic roots.
```
Comment (sidd-27) [inline crates/kornia-3d/src/ransac/estimators/ap3p.rs]: this calls solve_ap3p_multi which then allocates And returns an own vec which makes this contradictory to the zero heap allocation thing in the doc. Could you fix that?

File: crates/kornia-3d/src/pnp/ap3p.rs [hunk 3]
```diff
@@ -0,0 +1,944 @@
...
+        let rmse = rmse_px(&world_points, points_image, &r, &t, k);
+        if rmse < best_rmse {
+            best_rmse = rmse;
+            best_idx = Some(i);
+        }
+    }
+
+    let pick = best_idx.ok_or_else(|| {
+        PnPError::SvdFailed("AP3P: all candidates failed cheirality check".to_string())
+    })?;
```
Comment (sidd-27) [inline crates/kornia-3d/src/pnp/ap3p.rs]: I don't think the S V D failure here is appropriate.

File: crates/kornia-3d/src/pnp/ap3p.rs [hunk 4]
```diff
@@ -0,0 +1,944 @@
...
+    let mut feature_vectors = [[0.0f64; 3]; 3];
+    for i in 0..3 {
+        let u = points_image[i].x as f64;
+        let v = points_image[i].y as f64;
+        let mu = inv_fx * u - cx_fx;
+        let mv = inv_fy * v - cy_fy;
+        let norm = (mu * mu + mv * mv + 1.0).sqrt();
+        feature_vectors[i] = [mu / norm, mv / norm, 1.0 / norm];
+    }
+
```
Comment (sidd-27) [inline crates/kornia-3d/src/pnp/ap3p.rs]: The same setup happens byte for byte in the solve AP3P single function as well. Could this be deduplicated

**Comment by AdarshGuptaa** [comment]
Made the requested changes @sidd-27

**Review by edgarriba — APPROVED** [review]
